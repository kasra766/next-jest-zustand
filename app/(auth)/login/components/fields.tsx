"use client";
import { Button, Paper, Typography } from "@mui/material";
import { useForm } from "react-hook-form";
import { FormField } from "@/components/text-field";
import { useAuth } from "@/store/useAuth";

export function LoginFields() {
  const { login, signup } = useAuth();
  const { control, handleSubmit } = useForm({
    defaultValues: { user_name: "", password: "" },
  });

  const submit = (data: { user_name: string; password: string }) => {
    signup(data);
    login();
  };
  return (
    <Paper
      component="form"
      onSubmit={handleSubmit(submit)}
      elevation={12}
      className="flex min-w-[500px] max-w-[600px] flex-col gap-3 p-8"
    >
      <Typography variant="h4" mb={2} role="header">
        Login
      </Typography>

      <FormField
        control={control}
        name="user_name"
        rules={{ required: true }}
        type="text"
        autoFocus
        label="User Name"
        fullWidth
      />
      <FormField
        control={control}
        name="password"
        rules={{
          required: true,
          minLength: {
            value: 8,
            message: "password must be at least 8 character",
          },
        }}
        type="password"
        label="Password"
        fullWidth
      />
      <Button variant="contained" type="submit">
        submit
      </Button>
    </Paper>
  );
}
