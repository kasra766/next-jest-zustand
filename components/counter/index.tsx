"use client";
import { useCounter, useCounterStore } from "@/store/useCounter";
import { ButtonGroup, Button, Stack } from "@mui/material";

import AddIcon from "@mui/icons-material/Add";
import RemoveIcon from "@mui/icons-material/Remove";

export function Counter() {
  const { count, dec, inc } = useCounter();
  const increments = () => {
    inc();
  };
  return (
    <Stack justifyContent="center" alignItems={"center"} gap={3}>
      <p className="text-center" role="count_value">
        {count}
      </p>
      <ButtonGroup variant="contained">
        <Button
          onClick={inc}
          startIcon={<AddIcon />}
          variant="contained"
          data-testid="inc"
        >
          inc
        </Button>
        <Button
          onClick={dec}
          startIcon={<RemoveIcon />}
          variant="outlined"
          data-testid="dec"
        >
          dec
        </Button>
      </ButtonGroup>
    </Stack>
  );
}
