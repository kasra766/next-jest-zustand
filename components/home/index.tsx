"use client";
import { useAuth } from "@/store/useAuth";
import { Button, Paper, Stack, Typography } from "@mui/material";

export function Home() {
  const {
    user_information: { user_name },
    logout,
  } = useAuth();

  return (
    <Stack alignItems={"center"} pt={5}>
      <Paper
        elevation={12}
        className="flex min-w-[500px] max-w-[600px] flex-col gap-3 p-5"
      >
        <Typography
          variant="body1"
          textAlign={"center"}
          role="contentinfo"
        >{`welcome dear ${user_name}`}</Typography>
        <Button onClick={logout} variant="contained" role="button">
          Logout
        </Button>
      </Paper>
    </Stack>
  );
}
