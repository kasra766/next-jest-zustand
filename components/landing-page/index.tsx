"use client";
import Link from "next/link";
import { redirect } from "next/navigation";
import { Button, ButtonGroup, Stack, Typography } from "@mui/material";
import { useAuth } from "@/store/useAuth";

export function LandingPage() {
  const { isLogin } = useAuth();
  console.log(isLogin);

  if (isLogin) {
    redirect("/home");
  }
  return (
    <Stack justifyContent={"center"} alignItems={"center"} rowGap={3}>
      <Typography variant="h1">Welcome to over page</Typography>
      <Typography variant="body1">
        for seeing your profile please login or sign-up
      </Typography>
      <ButtonGroup size="large">
        <Button component={Link} href="/login" variant="contained">
          Login
        </Button>
        <Button component={Link} href="/sign-up" variant="outlined">
          Sign up
        </Button>
      </ButtonGroup>
    </Stack>
  );
}
