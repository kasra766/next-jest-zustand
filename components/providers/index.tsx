"use client";
import { SnackbarProvider } from "notistack";
import ThemeRegistry from "../theme-provider";

export function Providers({ children }: { children: React.ReactNode }) {
  return (
    <ThemeRegistry>
      <SnackbarProvider maxSnack={3}>{children}</SnackbarProvider>
    </ThemeRegistry>
  );
}
