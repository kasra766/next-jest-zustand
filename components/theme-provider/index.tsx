"use client";

import * as React from "react";
import { ThemeProvider } from "@mui/material/styles";
import CssBaseline from "@mui/material/CssBaseline";
import rtlPlugin from "stylis-plugin-rtl";
import NextAppDirEmotionCacheProvider from "./EmotionCache";
import theme from "./theme";

export default function ThemeRegistry({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <NextAppDirEmotionCacheProvider
      options={{ key: "mui", stylisPlugins: [rtlPlugin] }}
    >
      <ThemeProvider theme={theme}>
        {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
        <CssBaseline />
        {children}
      </ThemeProvider>
    </NextAppDirEmotionCacheProvider>
  );
}
