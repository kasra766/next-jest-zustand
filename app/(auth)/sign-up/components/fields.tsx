"use client";

import { useForm } from "react-hook-form";
import { Button, Paper, Stack, Typography } from "@mui/material";
import { FormField } from "@/components/text-field";
import { type UserSignup, useAuth } from "@/store/useAuth";
import { emailPattern, mobilePattern, strongRegex } from "@/lib/patterns";

export function SignUpFields() {
  const { userSignup: defaultValues, login, signup } = useAuth();
  const { control, handleSubmit } = useForm<UserSignup>({ defaultValues });

  const submit = (data: UserSignup) => {
    signup(data);
    login();
  };

  return (
    <Paper
      component="form"
      onSubmit={handleSubmit(submit)}
      elevation={12}
      className="flex min-w-[500px] max-w-[600px] flex-col  gap-3 p-8"
    >
      <Typography variant="h4" mb={2}>
        Sign Up
      </Typography>
      <Stack direction={"row"} columnGap={2}>
        <FormField
          control={control}
          name="fname"
          rules={{
            required: { value: true, message: "name is required" },
          }}
          type="text"
          label="Name"
          autoFocus
        />
        <FormField
          control={control}
          name="lname"
          rules={{
            required: { value: true, message: "last name is required" },
          }}
          type="text"
          label="Last Name"
        />
      </Stack>
      <Stack direction={"row"} columnGap={2}>
        <FormField
          control={control}
          name="phone"
          rules={{
            required: { value: true, message: "phone is required" },
            pattern: {
              value: mobilePattern,
              message: "mobiles format is wrong",
            },
          }}
          type="tel"
          label="Phone"
        />
        <FormField
          control={control}
          name="user_name"
          rules={{
            required: { value: true, message: "user name is required" },
          }}
          type="text"
          label="User Name"
        />
      </Stack>
      <FormField
        control={control}
        name="email"
        rules={{
          required: { value: true, message: "email is required" },
          pattern: {
            value: emailPattern,
            message: "emails format is wrong",
          },
        }}
        type="email"
        label="Email"
      />

      <FormField
        control={control}
        name="password"
        rules={{
          required: { value: true, message: "password is required" },
          minLength: { value: 8, message: "min length muse be 8 character" },
          pattern: {
            value: strongRegex,
            message:
              "password must consist at least an Uppercase and lowercase and special character(@,$,...) and numbers",
          },
        }}
        type="password"
        label="Password"
      />

      <Button fullWidth type="submit" variant="contained">
        Submit
      </Button>
    </Paper>
  );
}
