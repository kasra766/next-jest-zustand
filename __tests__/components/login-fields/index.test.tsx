import "@testing-library/jest-dom";
import "isomorphic-fetch";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { setupServer } from "msw/node";
import { rest } from "msw";
import { LoginFields } from "@/app/(auth)/login/components/fields";

const handlers = [
  rest.post("/api/signup-form", async (req, res, ctx) => {
    const { user_name } = await req.json();

    if (user_name === "amir") {
      return res(ctx.status(400), ctx.json({ message: "this user is ban" }));
    }
    return res(ctx.status(200), ctx.json({ message: "welcome" }));
  }),
];

const server = setupServer(...handlers);

describe("test login page", () => {
  const consoleSpy = jest.spyOn(console, "log").mockImplementation();
  beforeAll(() => server.listen());
  afterEach(() => {
    server.restoreHandlers();
    consoleSpy.mockClear();
    cleanup();
  });

  afterAll(() => server.close());
  it("check for header", () => {
    render(<LoginFields />);
    const header = screen.getByRole("header");
    expect(header).toHaveTextContent("Login");
  });
  it("check for submit button", () => {
    render(<LoginFields />);
    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("type", "submit");
  });
  it("check inputs label", () => {
    render(<LoginFields />);
    const input = screen.getByLabelText("User Name");
    expect(input).toHaveFocus();
  });

  describe("testing fields", () => {
    it("testing submit for status 200", async () => {
      render(<LoginFields />);
      fireEvent.change(screen.getByLabelText("User Name"), {
        target: {
          value: "kasra",
        },
      });

      fireEvent.change(screen.getByLabelText("Password"), {
        target: {
          value: "kasraK@1",
        },
      });

      fireEvent.submit(screen.getByTestId("submit-btn"));
      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith("success");
      });
    });

    it("testing submit for status 400", async () => {
      render(<LoginFields />);
      fireEvent.change(screen.getByLabelText("User Name"), {
        target: {
          value: "amir",
        },
      });

      fireEvent.change(screen.getByLabelText("Password"), {
        target: {
          value: "kasraK@1",
        },
      });

      fireEvent.submit(screen.getByTestId("submit-btn"));

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalledWith(
          "submit form err: something is wrong: 400",
        );
      });
    });
  });
});

describe("snap shot of login page", () => {
  it("snap shot", () => {
    const component = renderer.create(<LoginFields />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
      tree?.props.onSubmit();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
