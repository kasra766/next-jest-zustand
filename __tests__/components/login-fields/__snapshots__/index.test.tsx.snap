// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap shot of login page snap shot 1`] = `
<form
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation12 flex min-w-[500px] max-w-[600px] flex-col gap-3 p-8 css-15s4tt4-MuiPaper-root"
  onSubmit={[Function]}
>
  <h4
    className="MuiTypography-root MuiTypography-h4 css-pxw7rc-MuiTypography-root"
    role="header"
  >
    Login
  </h4>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r0:"
      id=":r0:-label"
    >
      User Name
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={true}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        name="user_name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            User Name
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r1:"
      id=":r1:-label"
    >
      Password
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r1:"
        name="password"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Password
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    submit
  </button>
</form>
`;

exports[`snap shot of login page snap shot 2`] = `
<form
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation12 flex min-w-[500px] max-w-[600px] flex-col gap-3 p-8 css-15s4tt4-MuiPaper-root"
  onSubmit={[Function]}
>
  <h4
    className="MuiTypography-root MuiTypography-h4 css-pxw7rc-MuiTypography-root"
    role="header"
  >
    Login
  </h4>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r0:"
      id=":r0:-label"
    >
      User Name
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={true}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        name="user_name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            User Name
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r1:"
      id=":r1:-label"
    >
      Password
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r1:"
        name="password"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="password"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Password
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    submit
  </button>
</form>
`;
